[git-webkit revert] Should error when radar is not found instead of creating bug with radar title

briannafan opened the pull request:
#### 0f11d38af6c8c8a2951e83ed944f4ffa282fb90c
<pre>
[git-webkit revert] Should error when radar is not found instead of creating bug with radar title
<a href="https://bugs.webkit.org/show_bug.cgi?id=268934">https://bugs.webkit.org/show_bug.cgi?id=268934</a>
<a href="https://rdar.apple.com/problem/122490424">rdar://problem/122490424</a>

Reviewed by Jonathan Bedard.

Returns an error when an issue is not found when a link is passed in.
Previous behaviour meant that the link would be used as a title to construct a new bug.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert):
(Revert.get_issue_info): Adds check so that radar and bugzilla links are not used as titles.
(Revert.create_revert_commit_msg): Consolidates issue logic.

Canonical link: <a href="https://commits.webkit.org/274885@main">https://commits.webkit.org/274885@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03529b385756470eb9ffdc77f5cff9fc5c743a6a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42654 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42821 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42583 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16617 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40850 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16238 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/40144 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44099 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/40323 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5330 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -42,12 +42,16 @@ class Revert(Command):
     REVERT_TITLE_TEMPLATE = 'Unreviewed, reverting {}'
     REVERT_TITLE_RE = re.compile(r'^Unreviewed, reverting {}'.format(Commit.IDENTIFIER_RE.pattern))
-    RES = [
+    RES_RADAR = [
```
I know this is old code, but this seems very wrong. This is duplicating regexes that are in `OpenSource/Tools/Scripts/libraries/webkitbugspy/webkitbugspy` already, tied to `Tracker` instances. So we shouldn't be re-declaring these here. Also, we shouldn't be trying to extract radars (or bugillas!) from repositories which don't have those trackers configured.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -100,6 +104,16 @@ def get_issue_info(cls, args, repository, commit_objects, commit_bugs, **kwargs)
         issue = Tracker.from_string(args.issue)
         commit_bugs_list = list(commit_bugs)
+        if not issue.title:
+            for regex in cls.RES_RADAR:
```
Can't we just check `issue.tracker`? Our `Tracker` objects should probably have a `name` variable (they don't right now), if they did, you could just have this be:
```
if not issue.title:
    sys.stderr.write('Could not fetch {} from link. Please verify that the issue exists.\n'.format(issue.tracker.name))
    return None
```

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -180,7 +194,7 @@ def create_revert_commit_msg(cls, args, commit_objects, **kwargs):
             for bug in CommitProgram.bug_urls(revert_issue):
-                for regex in cls.RES:
+                for regex in cls.RES_RADAR:
                     if regex.match(bug):
```
This seems like it should just be `Tracker.from_string()`, right? Since all of the regexes are centralized in the `Tracker` objects.

briannafan replied:
Good point! I also realized finding both links can be cut down to one iteration of `CommitProgram.bug_urls`.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -172,20 +170,11 @@ def create_revert_commit_msg(cls, args, commit_objects, **kwargs):
+            if isinstance(Tracker.from_string(issue_url).tracker, radar.Tracker):
+                revert_radar_link = issue_url
+            if isinstance(Tracker.from_string(issue_url).tracker, bugzilla.Tracker):
+                revert_bug_link = issue_url
```
This is fine for now, although we can generalize this too later. You aren't making this situation worse, though, so that shouldn't block landing this change.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
```diff
@@ -107,7 +107,7 @@ def radarclient():
     def __init__(self, users=None, authentication=None, project=None, projects=None, redact=None, hide_title=None, redact_exemption=None):
         hide_title = True if hide_title is None else hide_title
-        super(Tracker, self).__init__(users=users, redact=redact, redact_exemption=redact_exemption, hide_title=hide_title)
+        super(Tracker, self).__init__(type=self.NAME, users=users, redact=redact, redact_exemption=redact_exemption, hide_title=hide_title)
```
We already have `self.NAME`, we don't need `type`, right?

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -172,20 +170,11 @@ def create_revert_commit_msg(cls, args, commit_objects, **kwargs):
-                if revert_radar_link:
-                    break
+        for issue_url in CommitProgram.bug_urls(revert_issue):
+            if isinstance(Tracker.from_string(issue_url).tracker, radar.Tracker):
```
This is going to fall over if `Tracker.from_string()` returns `None`, will need to be this:
```
issue = Tracker.from_string(issue_url)
if not issue:
    continue
if isinstance(issue.tracker, radar.Tracker):
   ...
```

JonWBedard reviewed (approved):
r+ed, with a minor change

briannafan committed:
[git-webkit revert] Should error when radar is not found instead of creating bug with radar title
https://bugs.webkit.org/show_bug.cgi?id=268934
rdar://problem/122490424

Reviewed by Jonathan Bedard.

Returns an error when an issue is not found when a link is passed in.
Previous behaviour meant that the link would be used as a title to construct a new bug.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert):
(Revert.get_issue_info): Adds check so that radar and bugzilla links are not used as titles.
(Revert.create_revert_commit_msg): Consolidates issue logic.

Canonical link: https://commits.webkit.org/274885@main
